plugins/redis/redisasync.cpp

```diff
@@ -0,0 +1,870 @@
+            rtlFail(0, msg.str());
+        }
+    }
+}
+void AsyncConnection::connectCB(const redisAsyncContext * context, int status)
+{
+    if (status != REDIS_OK)
+    {
+        if (context->data)
+        {
+            const AsyncConnection * connection = (AsyncConnection *)context->data;
+            VStringBuffer msg("Redis Plugin : failed to connect to %s:%d - %s", connection->master.str(), connection->port, context->errstr);
+            rtlFail(0, msg.str());
+        }
+        else
+        {
+            VStringBuffer msg("Redis Plugin : failed to connect - %s", context->errstr);
+            rtlFail(0, msg.str());
+        }
+    }
+}
+void AsyncConnection::disconnectCB(const redisAsyncContext * context, int status)
+{
+    if (status != REDIS_OK)//&& context->err != 2)//err = 2:  ERR only (P)SUBSCRIBE / (P)UNSUBSCRIBE / QUIT allowed in this contex
+    {
+        if (context->data)
+        {
+            const AsyncConnection  * connection = (AsyncConnection*)context->data;
+            VStringBuffer msg("Redis Plugin : server (%s:%d) forced disconnect - %s", connection->master.str(), connection->port, context->errstr);
+            rtlFail(0, msg.str());
+        }
+        else
+        {
+            VStringBuffer msg("Redis Plugin : server forced disconnect - %s", context->errstr);
+            rtlFail(0, msg.str());
+        }
+    }
+}
+AsyncConnection * AsyncConnection::createConnection(ICodeContext * ctx, const char * options, unsigned __int64 database)
+{
+    return new AsyncConnection(ctx, options, database);
+}
+void AsyncConnection::assertRedisErr(int reply, const char * _msg)
+{
+    if (reply != REDIS_OK)
+    {
+        VStringBuffer msg("Redis Plugin: %s", _msg);
+        rtlFail(0, msg.str());
+    }
+}
+void AsyncConnection::assertOnError(const redisReply * reply, const char * _msg)
+{
+    if (!reply)
+    {
+        assertConnection();
+        //There should always be a connection error
+        VStringBuffer msg("Redis Plugin: %s%s", _msg, "no 'reply' nor connection error");
+        rtlFail(0, msg.str());
+    }
+    else if (reply->type == REDIS_REPLY_ERROR)
+    {
+        VStringBuffer msg("Redis Plugin: %s%s", _msg, reply->str);
+        rtlFail(0, msg.str());
+    }
+}
+void AsyncConnection::assertCallbackError(const redisAsyncContext * context, const redisReply * reply, const char * _msg)
+{
+    if (reply && reply->type == REDIS_REPLY_ERROR)
+    {
+        VStringBuffer msg("Redis Plugin: %s%s", _msg, reply->str);
+        rtlFail(0, msg.str());
+    }
+    assertContextErr(context);
+}
+void AsyncConnection::timeoutCB(struct ev_loop * evLoop, ev_timer *w, int revents)
+{
+
+    rtlFail(0, "Redis Plugin : async operation timed out");
+}
+void AsyncConnection::handleLoop(struct ev_loop * evLoop, ev_tstamp timeout)
+{
+    ev_timer timer;
+    ev_timer_init(&timer, timeoutCB, timeout, 0.);
+    ev_timer_again(evLoop, &timer);
+    ev_run(evLoop, 0);
```

Comment:
ev_run is the actual libev event loop. This is terminated by the timer (setup above) upon a timeout or when there are no more events open. In the case of this plugin such events are the hiredis libev adapter io read and write routines. These are closed within the relevant callbacks thus terminating the loop.